### PR TITLE
Allow scheduling with arbitrary user-defined resource labels.

### PR DESCRIPTION
**Pull request by robertnishihara**
This is intended to address #695.

This PR introduces the following API changes. Note that **all custom resources behave like GPUs**.

**Starting Ray from the command line.**

```
# The --resources argument is a JSON encoded dictionary.
ray start --head --num-cpus=64 --num-gpus=4 --resources='{"CustomLabel": 1}'
```

**Starting Ray from Python.**

```python
ray.init(num_cpus=64, num_gpus=4, resources={"CustomLabel": 1})
```

**Defining remote functions and actors.**

```python
@ray.remote(num_gpus=1, resources={"CustomLabel": 1})
def f():
    pass

@ray.remote(num_gpus=1, resources={"CustomLabel": 1})
class Actor(object):
    pass
```

**TODO in this PR:**
- [x] Add tests for the new functionality.
- [x] Fix remaining bugs.
- [x] Update all documentation.
- [x] Decide on API.
- [x] Add skipped test for actor placement with custom resources.
- [x] Add test to `multi_node_test.py`.
- [x] Double check handling when resource names are bytes versus unicode.
- [x] Maybe we really want GPU-like resources, not CPU-like resources. Figure out what to do about this.
- [x] Check for performance regressions.

**Not done in this PR:**
- Actor placement needs to be changed to take into account non-GPU resources.

### CODE REVIEW

**Comment by ericl** [comment]
It would be nice to retain backwards compatibility, and imo the original API read more nicely. How about something like:

`num_cpus=3, num_gpus=5, custom_resources={"foo": 5}`

**Comment by pcmoritz** [comment]
I agree with eric, we should have backwards compatibility/API niceness for the common case.

**Comment by robertnishihara** [comment]
I agree with API niceness, but it seems way too early to worry about backwards compatibility. I'd rather optimize for the best possible API right now even if that means breaking a few things.

File: src/common/lib/python/common_extension.cc [hunk 1]
```diff
@@ -323,20 +322,40 @@ static int PyTask_init(PyTask *self, PyObject *args, PyObject *kwds) {
...
+                        "the keys in resource_map must be strings");
+        return -1;
+      }
+      if (!(PyFloat_Check(value) || PyLong_Check(value))) {
+        PyErr_SetString(PyExc_TypeError,
+                        "the values in resource_map must be floats");
+        return -1;
+      }
+
+      std::string resource_name(PyBytes_AsString(PyUnicode_AsASCIIString(key)),
```
Comment (pcmoritz) [inline src/common/lib/python/common_extension.cc]: Careful here! PyUnicode_AsASCIIString creates a new reference and needs to be decrefed.

**Comment by aldanor** [comment]
So, `'CPU'` and `'GPU'` are going to be special resource strings?

Also wondering how does this play with "infinite resources" (i.e., "labels")?

**Comment by atumanov** [comment]
@aldanor , this is a good question. The intention was to subsume both finite capacity and infinite capacity resources as "labels". It should be straightforward to set label capacity to infinity and perform arithmetic on that  obliviously.  I would much rather not special case labels as separate from resources, if at all possible.

**Comment by atumanov** [comment]
@aldanor , regarding your first question, this PR is a first step towards proper label support. I think strings are fine for now, as this is strictly more expressive than having predefined resource label constants. Label management (e.g., YARN-style) is a whole different dimension that we have to decide whether we want to tackle going forward. If you have thoughts on that, let us know.

File: src/common/common_protocol.cc [hunk 2]
```diff
@@ -24,3 +24,30 @@ to_flatbuf(flatbuffers::FlatBufferBuilder &fbb,
   }
   return fbb.CreateVector(results);
 }
+
+std::string string_from_flatbuf(const flatbuffers::String *string) {
+  return std::string(string->data(), string->size());
+}
+
+const std::unordered_map<std::string, double> map_from_flatbuf(
+    const flatbuffers::Vector<flatbuffers::Offset<ResourcePair>>
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply src/common/common_protocol.cc]: Sure, I copied it from this line 

https://github.com/ray-project/ray/blob/ad044cbe8f9b39ffbfd12ba68af2b88986eda96c/src/common/common_protocol.cc#L9

I'll change it there as well.

File: src/common/lib/python/common_extension.cc (hunk 1 above)
Comment (robertnishihara) [reply src/common/lib/python/common_extension.cc]: Good catch!

**Comment by robertnishihara** [comment]
Sticking with `num_cpus`, `num_gpus` for CPUs and GPUs for now. We can change this in a follow up PR if desired..

**Comment by robertnishihara** [comment]
There is a non-trivial performance slowdown when using very large numbers of custom resources.

**No custom resources**

```python
import ray
ray.init()

@ray.remote
def f():
    return 1

%time l = ray.get([f.remote() for _ in range(10000)])  # 950 ms
```

**Many custom resources**

```python
import ray
ray.init(resources={str(i): 100 for i in range(10000)})

@ray.remote
def f():
    return 1

@ray.remote(resources={str(i): 1 for i in range(10000)})
def g():
    return 1

%time l = ray.get([f.remote() for _ in range(10000)])  # 17.1 s
%time l = ray.get([g.remote() for _ in range(10000)])  # 2min 26s
```

**Comment by robertnishihara** [comment]
After simplifying the checks it is

```python
%time l = ray.get([f.remote() for _ in range(10000)])  # 921 ms
%time l = ray.get([g.remote() for _ in range(10000)])  # 1.43 s
%time l = ray.get([h.remote() for _ in range(10000)])  # 2min 31s
```

The remaining slowdown for `g.remote()` is due to the fact that `resources_available` in `local_scheduler_algorithm.cc` takes time linear in the number of custom resources the local scheduler is aware of. Removing that function takes it from **1.43s** to **954ms**.

**Comment by aldanor** [comment]
> @aldanor , regarding your first question, this PR is a first step towards proper label support. I think strings are fine for now, as this is strictly more expressive than having predefined resource label constants. Label management (e.g., YARN-style) is a whole different dimension that we have to decide whether we want to tackle going forward. If you have thoughts on that, let us know.

Thanks for clarifying -- that makes sense.

I also wanted to ask whether dynamic resource constraints would be feasible to support within the current/future framework? E.g., even with just the core resources, say there's a remote task that takes one array input and consumes 3x its size in RAM while executing -- that's the knowledge we have upfront but it will only become a concrete number at runtime when the shape of the input is known, at some point during the execution.

**Comment by robertnishihara** [comment]
@aldanor I think the easiest way to do this would be to have the local scheduler on a given machine probe the machine's resource utilization (CPU, memory, etc) and decide to schedule more or fewer tasks based on this information.

Since tasks can be arbitrary Python functions, it would be a little difficult to inspect the function and figure out it's actual resource requirements. There might be a good way to allow hints to the scheduler about this kind of thing, though I'm not quite sure what the API would look like.

**Comment by atumanov** [comment]
@aldanor , there are two sides to your question w.r.t. support for dynamic resource requirements: accounting for them and enforcing them. Accounting should be straightforward. This basically involves changing the Task's resource requirement vector at runtime. I don't see any fundamental difficulty with doing that. Once changed (e.g., updating MEMREQ from, say 128MB to 512MB), the task will be scheduled in accordance with its latest known task spec. The second part, enforcement is orthogonal and is a little more involved. In the simplest case, it would require python workers to set rlimits on memory for each task that comes with such a memory limit requirement. 
Question is, can you think of any use cases other than memory where updating resource requirements dynamically would make sense?

**Comment by robertnishihara** [comment]
@pcmoritz @atumanov I've rebased this and made a few fixes. Should be ready to review/merge pending tests passing.

**Comment by aldanor** [comment]
@atumanov I'm not sure about other cases, but RAM usage in a lot of cases is runtime-known (and CPU usage too, in some cases). For example, you have a remote task that receives information on where to read the data and which learner to run on it; as such, you know exactly the RAM usage, but it's not a property of the task itself, but rather of a specific call -- and then you schedule a whole grid of these tasks. I didn't quite get the point about task-specific rlimits though as this resource usage would still fall within node's rlimit?

**Comment by robertnishihara** [comment]
@aldanor would the memory requirement be known at task *submission* time or task *execution* time? If the information isn't available until *execution* time then it will be a little harder to make use of. If it's known at *submission* time, then we can definitely make use of the information. Do you have any thoughts about what the API might look like?

File: test/runtest.py [hunk 3]
```diff
@@ -1495,19 +1495,87 @@ def h():
...
+        # Make sure that tasks with unsatisfied custom resource requirements do
+        # not get scheduled.
+        ready_ids, remaining_ids = ray.wait([j.remote(), k.remote()],
+                                            timeout=500)
+        self.assertEqual(ready_ids, [])
+
+        ray.worker.cleanup()
+
+    def testManyCustomResources(self):
+        num_custom_resources = 10000
```
Comment (robertnishihara) [inline test/runtest.py]: Note that if you make this number very large (e.g., 10x bigger), the command that we use to start the local scheduler from the command line is extremely long and will fail.

File: src/common/lib/python/common_extension.cc [hunk 4]
```diff
@@ -410,10 +442,20 @@ static PyObject *PyTask_arguments(PyObject *self) {
...
   TaskSpec *task = ((PyTask *) self)->spec;
-  PyObject *required_resources = PyList_New((Py_ssize_t) ResourceIndex_MAX);
-  for (int i = 0; i < ResourceIndex_MAX; ++i) {
-    double r = TaskSpec_get_required_resource(task, i);
-    PyList_SetItem(required_resources, i, PyFloat_FromDouble(r));
+  PyObject *required_resources = PyDict_New();
+  for (auto const &resource_pair : TaskSpec_get_required_resources(task)) {
+    std::string resource_name = resource_pair.first;
+#if PY_MAJOR_VERSION >= 3
+    PyObject *key = PyUnicode_FromStringAndSize(resource_name.c_str(),
```
Comment (pcmoritz) [inline src/common/lib/python/common_extension.cc]: .data() is better here

File: src/common/lib/python/common_extension.cc [hunk 5]
```diff
@@ -410,10 +442,20 @@ static PyObject *PyTask_arguments(PyObject *self) {
...
+                                                resource_name.size());
+#else
+    PyObject *key =
+        PyBytes_FromStringAndSize(resource_name.c_str(), resource_name.size());
```
Comment (pcmoritz) [inline src/common/lib/python/common_extension.cc]: .data() is better here

File: src/local_scheduler/local_scheduler.cc [hunk 6]
```diff
@@ -412,94 +412,106 @@ LocalSchedulerState *LocalSchedulerState_init(
...
-      state->dynamic_resources[ResourceIndex_CustomResource] <
-          num_custom_resource) {
-    return false;
-  }
-  if (state->dynamic_resources[ResourceIndex_GPU] < num_gpus) {
-    return false;
+bool check_dynamic_resources(
+    LocalSchedulerState *state,
+    const std::unordered_map<std::string, double> &resources) {
+  for (auto const &resource_pair : resources) {
```
Comment (robertnishihara) [inline src/local_scheduler/local_scheduler.cc]: There's a slight change in intended usage here. Before, we only checked `CPU` resources if the passed in quantity was `> 0`. Now if we don't want to check `CPU` resources, then `CPU` should simply not be in `resources`.

**Comment by aldanor** [comment]
> @aldanor would the memory requirement be known at task submission time or task execution time? If the information isn't available until execution time then it will be a little harder to make use of. If it's known at submission time, then we can definitely make use of the information. Do you have any thoughts about what the API might look like?

@robertnishihara I surely meant *submission* time, this will cover most common use cases I believe; but it will be already a huge win compared to static rlimits. Execution time rlimits would indeed be a much trickier task and I off top of my head I can't find an immediate use case for it.

Not sure about the API; because kwargs are out of question if they are passed through, it has to be something else, like an external function or maybe an attribute. Some bikeshedding (where `rlimit` could be something nicer like `with_resources` or whatever):

```python
f.remote(x=1).rlimit(num_cpus=2, resources={'MyResource': 3})  # ?

rlimit(f.remote(x=1), num_cpus=2, resources={'MyResource': 3})  # ?
```

File: src/common/test/db_tests.cc [hunk 7]
```diff
@@ -68,13 +68,17 @@ int64_t timeout_handler(event_loop *loop, int64_t id, void *context) {
 TEST object_table_lookup_test(void) {
   event_loop *loop = event_loop_create();
   /* This uses manager_port1. */
-  const char *db_connect_args1[] = {"manager_address", "127.0.0.1:12345"};
+  std::vector<std::string> db_connect_args1;
```
Comment (pcmoritz) [inline src/common/test/db_tests.cc]: an alternative here is to use `std::vector<std::string> db_connect_args1 = {"manager_address", "127.0.0.1:12345"}`

**Comment by robertnishihara** [comment]
@aldanor, the underlying infrastructure is in place to specify resource requirements at task submission time. You're right that there is no real need to require all tasks that execute a given remote function to have the same resources requirements. So the current API is too inflexible.

I think the only thing preventing this feature is the lack of a clean API to specify resource requirements when submitting tasks, but I think you're right it would be useful. One workaround for now could be to just define multiple remote functions with different resource requirements. E.g.,

```python
def f():
    pass

f1 = ray.remote(resources={"Memory": 1})(f)
f2 = ray.remote(resources={"Memory": 2})(f)
f3 = ray.remote(resources={"Memory": 3})(f)
```

and then call whichever is appropriate.

Want to create an issue so we can keep track of this issue separately from this PR?

**Comment by atumanov** [comment]
@aldanor , your earlier comment posted today clarified the use case, and, yes, what you want is tasks *submission* time resource requirement specification.

**Comment by aldanor** [comment]
@robertnishihara Done, I've opened #1278. (btw it would simplify some of the tests in this PR as well)

**Comment by robertnishihara** [comment]
@aldanor thanks, you're right about that!

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**